### ISSUE TITLE
luigi.contrib.sqla - Boolean values now cause a type error

### ISSUE THREAD

**Issue by andyalky** [spotify/luigi#2347]
I have a script that, as part of my data pipeline, loads a TSV file to a database. This is done through the luigi.contrib.sqla module.

SQLAlchemy released version 1.2 on January 24th. When I upgraded SQLAlchemy from version 1.1 to version 1.2, I started getting the following errors when running my script:

```
File "[...]/env/lib/python3.6/site-packages/sqlalchemy/sql/sqltypes.py", line 1653, in _strict_as_bool
    "Not a boolean value: %r" % value)
TypeError: Not a boolean value: 'True'
```

My database contains Boolean values, and previously, all my Boolean values were represented as "True" or "False" in my TSV input files. I tried changing the values in the TSV to "0" and "1", yet the error persisted. 

I checked SQLAlchemy's release notes, and the new release mandated that Boolean values no longer be passed through as strings.

According to their documentation:

> "The ultimate solution to this problem is that **string values are not supported with Boolean**, so in 1.2 a hard TypeError is raised if a non-integer / True/False/None value is passed. Additionally, only the integer values 0 and 1 are accepted."
(http://docs.sqlalchemy.org/en/latest/changelog/migration_12.html#boolean-datatype-now-enforces-strict-true-false-none-values)

When I downgrade my SQLAlchemy to 1.1, my script works as intended.

SQLAlchemy's release notes propose a fix by using SQLAlchemy's TypeDecorator. I'd like to try implementing this, but I'm not sure if there is a better/easier way to fix this issue for Luigi.

Would I be correct in suspecting that, due to SQLAlchemy's recent release, luigi.contrib.sqla is no longer able to handle Boolean values as expected?

Thanks!

**Comment by dlstadther** [comment]
@andyalky Feel free to contribute back to Luigi's `contrib.sqla` module to handle compatibility for both SQLAlchemy <=1.1. and >=1.2.

**Comment by gitlines** [comment]
Using SQLAlchemy==1.2.7 and Flask-SQLAlchemy==2.3.2 and a PostgreSQL database I hit the same problem - even if I change the value to integers like 0 or 1. How does one implement boolean values now?

**Comment by jerryji** [comment]
Google led me here as I was struggling with the following error in my Flask application --

sqlalchemy.exc.StatementError: (builtins.TypeError) Not a boolean value: False [SQL: 'INSERT INTO ... '] [parameters: [{'my_boolean': (False,), ...]

Very similar symptom: the error only happens in SQLAlchemy 1.2.X and goes away in SQLAlchemy 1.1.18. After hours of futile effort of trying to replicate the problem for an SQLAlchemy bug report, I finally noticed that it's all caused by a stupid mistake in my model class --

```
def __init__(self, my_boolean):
    self.my_boolean = my_boolean,
```
 
Notice the problem? There's a trailing comma at the end of my assignment.

Though my issue and solution is not related to luigi, I do hope my experience could help one way or the other.

**Comment by muhammadmohsinnbs** [comment]
Hi Guys,
I'm facing following error in SQLAlchemy ORM while inserting data into postgresql. The value for Boolean I'm providing is "false" or "true" which is acceptable by postgresql. I tried to insert data manually via sqlworkbench and its working alright but getting exception in SqlAlchemy:

Exception: (builtins.TypeError) Not a boolean value: 'true'

Can anyone please help me in this, am I doing something wrong?

**Comment by kevinsqi** [comment]
This change in the sqlalchemy 1.2 changelog may be relevant: [Boolean datatype now enforces strict True/False/None values](https://docs.sqlalchemy.org/en/13/changelog/migration_12.html#change-4102).

> In version 1.1, the change described in Non-native boolean integer values coerced to zero/one/None in all cases produced an unintended side effect of altering the way Boolean behaves when presented with a non-integer value, such as a string. In particular, the string value "0", which would previously result in the value False being generated, would now produce True. Making matters worse, the change in behavior was only for some backends and not others, meaning code that sends string "0" values to Boolean would break inconsistently across backends.

> The ultimate solution to this problem is that string values are not supported with Boolean, so in 1.2 a hard TypeError is raised if a non-integer / True/False/None value is passed. Additionally, only the integer values 0 and 1 are accepted.

> To accommodate for applications that wish to have more liberal interpretation of boolean values, the TypeDecorator should be used. Below illustrates a recipe that will allow for the “liberal” behavior of the pre-1.1 Boolean datatype:

```python
from sqlalchemy import Boolean
from sqlalchemy import TypeDecorator

class LiberalBoolean(TypeDecorator):
    impl = Boolean

    def process_bind_param(self, value, dialect):
        if value is not None:
            value = bool(int(value))
        return value
```

**Comment by kirankumar-thotta** [comment]
I see same error when I try to insert a boolean value to PostgreSQL DB table through Web form, can you please let me if there is any work around for this issue.

sqlalchemy.exc.StatementError: (builtins.TypeError) Not a boolean value: 'True'

Version i'm on 
SQLAlchemy       1.3.1
Flask-Migrate    2.5.3
Python 3.8.2

Thanks

**Comment by passitalong** [comment]
I also ran into this error. The code provided by kevinsql above was the key to our solution. It seems bizarre that an ORM can't handle Booleans as we verified we were sending actual bool's but they were being converted to tuples.

Not sure if it is SQLAlchemy or Flask-SQLAlchemy but some values are being transformed into single length tuples. So we had to modify the code to something like:

`class LiberalBoolean(TypeDecorator):
    impl = Boolean

    def process_bind_param(self, value, dialect):

        if value is not None:
            if isinstance(value, tuple):
                value = value[0]
            if isinstance(value,bool):
                return value
            value = bool(int(value))
        return value`

We are using:
Flask==2.0.2
Flask-SQLAlchemy==2.5.1
psycopg2-binary==2.9.1
SQLAlchemy==1.4.26
